flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCheckpointPlanCalculator.java

```diff
@@ -0,0 +1,329 @@
+                                        + "Aborting checkpoint.",
+                                execution.getVertex().getTaskNameWithSubtaskIndex(), jobId),
+                        CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+            }
+        }
+    }
+
+    /**
+     * Computes the checkpoint plan when all tasks are running. It would simply marks all the source
+     * tasks as need to trigger and all the tasks as need to wait and commit.
+     *
+     * @return The plan of this checkpoint.
+     */
+    private CheckpointPlan calculateWithAllTasksRunning() {
+        List<Execution> executionsToTrigger =
+                sourceTasks.stream()
+                        .map(ExecutionVertex::getCurrentExecutionAttempt)
+                        .collect(Collectors.toList());
+
+        List<Execution> tasksToWaitFor = createTaskToWaitFor(allTasks);
+
+        return new CheckpointPlan(
+                Collections.unmodifiableList(executionsToTrigger),
+                Collections.unmodifiableList(tasksToWaitFor),
+                Collections.unmodifiableList(allTasks),
+                Collections.emptyList(),
+                Collections.emptyList());
+    }
+
+    /**
+     * Calculates the checkpoint plan after some tasks have finished. We iterate the job graph to
+     * find the task that is still running, but do not has precedent running tasks.
+     *
+     * @return The plan of this checkpoint.
+     */
+    private CheckpointPlan calculateAfterTasksFinished() {
+        Map<JobVertexID, Integer> numberOfRunningTasksByVertex = countRunningTasks();
+
+        List<Execution> tasksToTrigger = new ArrayList<>();
+
+        List<Execution> tasksToWaitFor = new ArrayList<>();
+        List<Execution> finishedTasks = new ArrayList<>();
+        List<ExecutionJobVertex> fullyFinishedJobVertex = new ArrayList<>();
+
+        for (ExecutionJobVertex jobVertex : jobVerticesInTopologyOrder) {
+            int numberOfRunningTasks = numberOfRunningTasksByVertex.get(jobVertex.getJobVertexId());
+
+            if (numberOfRunningTasks == 0) {
+                fullyFinishedJobVertex.add(jobVertex);
+
+                for (ExecutionVertex task : jobVertex.getTaskVertices()) {
+                    finishedTasks.add(task.getCurrentExecutionAttempt());
+                }
+
+                continue;
+            }
+
+            List<JobEdge> prevJobEdges = jobVertex.getJobVertex().getInputs();
+
+            // this is an optimization: we determine at the JobVertex level if some tasks can even
+            // be eligible for being in the "triggerTo" set.
+            boolean someTasksMustBeTriggered =
+                    someTasksMustBeTriggered(numberOfRunningTasksByVertex, prevJobEdges);
+
+            for (ExecutionVertex vertex : jobVertex.getTaskVertices()) {
+                if (!vertex.getCurrentExecutionAttempt().isFinished()) {
+                    tasksToWaitFor.add(vertex.getCurrentExecutionAttempt());
```

Comment:
Very thanks for the comment! It should indeed more effective since in `hasRunningPrecedentTasks` we may need to access the `isFinished` for multiple times. According to the previous tests, I recorded the running status in `BitSet`, it should be able to compress the memory used and also avoid the time used for dynamic memory allocation~

Reply:
Thanks for updating, I see `hasRunningPrecedentTasks` now doesn't access `volatile` field.
However, 2nd access is still there (`calculateAfterTasksFinished`).

Reply:
Very sorry for missing change this part... I fixed this issue: https://github.com/apache/flink/pull/14740/commits/3127cc7d9cb2eed93319e62b13e6b6b13175b82b

Reply:
No worries, thanks for fixing this.